search/floydwarshall.go

```diff
@@ -156,38 +157,62 @@ func (p ShortestPaths) Between(u, v graph.Node) (path []graph.Node, weight float
 	}
 	path = []graph.Node{p.nodes[from]}
 	unique = true
+	seen := make([]int, len(p.nodes))
+	for i := range seen {
+		seen[i] = -1
+	}
+	seen[from] = 0
 	for from != to {
 		c := p.at(from, to)
 		if len(c) != 1 {
 			unique = false
 		}
-		from = c[rand.Intn(len(c))]
+		i := rand.Intn(len(c))
+		from = c[i]
+		if seen[from] >= 0 {
```

Comment:
> Yes, that's what I meant. Sorry.

In fact, that's not what I meant :-) I thought that the path is trimmed back to the beginning of the cycle, which was a ridiculous thought.